### ISSUE TITLE
table {x,y} resolving on draw

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#20068]
Trying to get the x and y coordinate of each cell, the get_y method returns the same value
```python
fig, ax = plt.subplots()
tab = mtable.table(ax, ['1','2', '3'])
for i in range(3):
    print(tab[i,0].get_y())
```
prints
```
0
0
0
```
but putting the loop in its own cell (w/o the fig, ax) yields the correct values (even though it's the same object as above)
```python
for i in range(3):
    print(tab[i,0].get_y())
```
```
-0.05518763796909486
-0.11037527593818977
-0.1655629139072847
```
a .py file also prints the incorrect values for the table cell x values. I've tried changing the variable name and got the same error in [colab](https://colab.research.google.com/drive/1YwkatG5311LaEFhB2AevXZLnq5sJYwXu?usp=sharing).

**Comment by jklymak** [comment]
I imagine you need to draw first.

**Comment by story645** [comment]
The problem is the coordinates don't resolve consistently at draw time when combined with ConnectionPatch.

When explicitly accessing the table elements, in a map from one axes to another, the image that's produced has all the arrows at the same y (which it currently sets to 0) when the should be at different ys
### incorrect
 ![image](https://user-images.githubusercontent.com/1300499/116008638-9de32380-a5e3-11eb-8dc5-e31e825c9519.png)
### correct
![image](https://user-images.githubusercontent.com/1300499/116008771-38436700-a5e4-11eb-847b-e8530dc88443.png)

```python
import matplotlib.patches as mpatches
import matplotlib.table as mtable

fig, (ax1, ax2) = plt.subplots(ncols=2)

tt1 = mtable.table(ax1, ['1','2', '3'])
tt2 = mtable.table(ax2, ['4', '5', '6'])
ax1.axis('off')
ax2.axis('off')

x01, y01, = tt1[0,0].get_x(), tt1[0,0].get_y()
x02, y02 = tt2[0,0].get_x(), tt2[0,0].get_y()
c1 = mpatches.ConnectionPatch(xyA=(x01,y01), coordsA=ax1.transData, 
                          xyB=(x02, y02), coordsB=ax2.transData, 
                              color='tab:orange', lw=2)
fig.add_artist(c1)

x11, y11, = tt1[1,0].get_x(), tt1[1,0].get_y()
x12, y12 = tt2[1,0].get_x(), tt2[1,0].get_y()
c2 = mpatches.ConnectionPatch(xyA=(x11,y11), coordsA=ax1.transData, 
                          xyB=(x12,y12), coordsB=ax2.transData, 
                              color='tab:orange', lw=2)
fig.add_artist(c2)
```

**Comment by jklymak** [comment]
Calling `fig.draw_no_output()` or `fig.canvas.draw()` after you make the tables but before you call `get_y` fixes this problem.  The table needs to be rendered before it knows how big its own cells are.  I'm going to close because I don't think there is an action item here.

**Comment by jklymak** [comment]
... if you are going to re-open, can you make it clear what the next step is?

**Comment by story645** [comment]
I think it should work w/o having to explicitly call `fig.canvas.draw()` especially here where it's passed into an object that also will not be resolved 'til draw time. 

Also I'm still confused 'cause I stumbled on this when trying to implement behavior that worked fine when making a different visualization [code here](https://github.com/story645/proposal/blob/055f117e999c8ff4f0c4ae90b0c61b5edc476113/figures/intro/viz_start.ipynb), and I couldn't reproduce why it worked. It either consistently should or shouldn't. 
![image](https://user-images.githubusercontent.com/1300499/116009469-ebfa2600-a5e7-11eb-8664-0837a9471139.png)

**Comment by jklymak** [comment]
Yeah I really don't think we can do anything here.  The table needs to know how big the text is and that requires a draw.

**Comment by timhoffm** [comment]
The desire that it should just work is understandable. However, I don't think we can do anything here (other than force a draw when creating a table, which would be against our usual policy and could potentially have other unwanted side effects).

**Comment by jklymak** [comment]
Well, we could force a draw on querrying x or y if the table is stale.  I *think* we do that somewhere, and maybe could start in other places.  But the draw still has to be global, which is expensive.

**Comment by story645** [comment]
get_x/get_y should resolve to the right variables when connection patch is being drawn (so at draw time anyway...) - is there a way to force the ordering or something?

**Comment by timhoffm** [comment]
Many positions are not correct until a draw has been performed, e.g.
~~~
In [1]: import matplotlib.pyplot as plt

In [2]: t = plt.xlabel('test')

In [3]: t.get_position()
Out[3]: (0.5, 0)

In [4]: plt.draw()

In [5]: t.get_position()
Out[5]: (0.5, 23.52222222222222)
~~~

This is a fundamental limitation in the current design of matplotlib. I'd go as far as "You cannot trust any position before the plot is drawn."

So for your case, my recommendation is be to manually insert a draw. Unfortunately, there's no free lunch here.

**Comment by jklymak** [comment]
> is there a way to force the ordering or something?

I don't think there is anything you can change or force.  You want to know x and y before _any_ artists are drawn, and that is unfortunately not possible.  

As a showcase of Matplotlib's capabilities, what you are doing is great, if perhaps more efficiently done by hand in a graphics package.  But if you are going to use complicated interactions that depend on the position of text on the figure, you are going to have to expect to occasionally pre-render.

**Comment by tacaswell** [comment]
Probably what we need to add here is the ability for `ConnectionPatch` (and depending on how far we want to pushing _most_ things that can take a transform)  to take an `Artist` for `coordsA` and `coordsB` like `annotation` can.  This way we can defer resolving the location of the Aritst until draw time and then interpret the coordinate as being normalized on the unit square of the Artists bounding box.  The code that does this is `Annotation` is: https://github.com/matplotlib/matplotlib/blob/178b46ebf780eaa07a7a732cde1f2795f4ebce78/lib/matplotlib/text.py#L1436-L1438

Another maybe better (but also one I do not know how to do off the top of my head) option is to add a transform to the `Artist` baseclass that will give an automatically updating transform that tracks its bounding box.

**Comment by timhoffm** [comment]
> to take an `Artist`

probably better a callable. There are too many degrees of freedom here (what if you want the `ConnectionPatch` vertically centered at the side of the cell). Not sure if we should go down this road though, looks like a big rabbit hole; e.g. evaluation order is important.
Also this does not feel like a very intuitive API.

Another alternative would be a completely new "layout" mechanism, similar to GUIs where you specify relations rather than positions (like .e.g. a layout grid and alignments within that grid). But that'd be a major project.

**Comment by anntzer** [comment]
FWIW there's already precedent for specifying coordinates in relative units of the bbox of another artist (`Annotation(..., xy=<somecoords>, xycoords=<someartist>)`; likewise `image.BboxImage` can be thought of as offering a special-case of that.  So it would be nice if these could be made more general.

But I agree that there are evaluation order problems here; e.g. I guess you could in theory end up with a loop of annotations refering to each other's bboxes and get something completely wrong, but I don't think(?) we want to go the route of trying to detect such loops either :/

**Comment by jklymak** [comment]
I guess it'd be possible to add an `add_annotation` method that kept a second annotation queue that was meant to happen after the other draws.

**Comment by timhoffm** [comment]
> I guess it'd be possible to add an `add_annotation` method that kept a second annotation queue that was meant to happen after the other draws.

Um, maybe I'm reading this wrong, but ad-hoc addition of additional drawing queues and related API is not a sustainable way.

**Comment by timhoffm** [comment]
Per the discussion above, we unfortunately cannot fix this in a reasonable way.

**Comment by story645** [comment]
> we unfortunately cannot fix this in a reasonable way.

I think the general fix is #22223, but I'm fine with this issue being closed in favor of that one.
